apollo-relay/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+        let header = Some(ReportHeader::try_new(graph)?);
+
+        Ok(Report {
+            header,
+            ..Default::default()
+        })
+    }
+}
+
+#[cfg(target_os = "windows")]
+fn get_uname() -> Result<String, std::io::Error> {
+    // Best we can do on windows right now
+    let sysname = sys_info::os_type().unwrap_or("Windows".to_string());
+    let nodename = sys_info::hostname().unwrap_or("unknown".to_string());
+    let release = sys_info::os_release().unwrap_or("unknown".to_string());
+    let version = "unknown";
+    let machine = "unknown";
+    Ok(format!(
+        "{}, {}, {}, {}, {}",
+        sysname, nodename, release, version, machine
+    ))
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_uname() -> Result<String, std::io::Error> {
+    let u = uname::uname()?;
+    Ok(format!(
+        "{}, {}, {}, {}, {},",
+        u.sysname, u.nodename, u.release, u.version, u.machine
+    ))
+}
+
+impl ReportHeader {
+    fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let mut header = ReportHeader {
+            ..Default::default()
+        };
+
+        header.add_hostname()?;
+        header.agent_version = std::env!("CARGO_PKG_VERSION").to_string();
+        header.runtime_version = "N/A".to_string();
+        header.uname = get_uname()?;
+        header.graph_ref = graph.to_string();
+        Ok(header)
+    }
+
+    fn add_hostname(&mut self) -> Result<(), ReporterError> {
+        self.hostname = hostname()?;
+        Ok(())
+    }
+}
+
+/// The Reporter accepts requests from clients to transfer statistics
+/// and traces to the Apollo Ingress relay.
+#[derive(Debug)]
+pub struct Reporter {
+    client: ReporterClient<Channel>,
+}
+
+impl Reporter {
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new<T: AsRef<str>>(addr: T) -> Result<Self, ReporterError>
+    where
+        prost::bytes::Bytes: From<T>,
+    {
+        let ep = Endpoint::from_shared(addr)?;
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new_with_static(addr: &'static str) -> Result<Self, ReporterError> {
+        let ep = Endpoint::from_static(addr);
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Submit these stats onto the relayer for eventual processing.
+    ///
+    /// The relayer will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_stats(
+        &mut self,
+        graph: ReporterGraph,
+        q: String,
```

Comment:
The term we use for this particular string is "stats report key" (which I guess is a bit outdated since the report has more than just stats, but meh). I think it would be more clear if you use that term consistently as well.

Reply:
ok